starky2/src/cross_table_lookup.rs

```diff
@@ -0,0 +1,294 @@
+                        next_z: *looking_z_next,
+                        challenges,
+                        columns: looking_columns,
+                    });
+
+                    let (looked_z, looked_z_next) = ctl_zs[*looked_table as usize].next().unwrap();
+                    acc[*looked_table as usize].push(Self {
+                        local_z: *looked_z,
+                        next_z: *looked_z_next,
+                        challenges,
+                        columns: looked_columns,
+                    });
+                }
+                acc
+            })
+    }
+}
+
+pub(crate) fn eval_cross_table_lookup_checks<F, FE, P, C, S, const D: usize, const D2: usize>(
+    vars: StarkEvaluationVars<FE, P, { S::COLUMNS }, { S::PUBLIC_INPUTS }>,
+    ctl_vars: &[CtlCheckVars<F, FE, P, D2>],
+    consumer: &mut ConstraintConsumer<P>,
+) where
+    F: RichField + Extendable<D>,
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+    C: GenericConfig<D, F = F>,
+    S: Stark<F, D>,
+{
+    for lookup_vars in ctl_vars {
+        let CtlCheckVars {
+            local_z,
+            next_z,
+            challenges,
+            columns,
+        } = lookup_vars;
+        let mut factor = ReducingFactor::new(challenges.beta);
+        let mut combine = |v: &[P]| -> P {
+            factor.reduce_ext(columns.iter().map(|&i| v[i])) + FE::from_basefield(challenges.gamma)
+        };
+
+        // Check value of `Z(1)`
+        consumer.constraint_first_row(*local_z - combine(vars.local_values));
+        // Check `Z(gw) = combination * Z(w)`
+        consumer.constraint_transition(*next_z - *local_z * combine(vars.next_values));
+    }
+}
+
+pub(crate) fn verify_cross_table_lookups<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    cross_table_lookups: Vec<CrossTableLookup>,
+    proofs: &[&StarkProofWithPublicInputs<F, C, D>],
+    challenges: GrandProductChallengeSet<F>,
+    config: &StarkConfig,
+) -> Result<()> {
+    let degrees_bits = proofs
+        .iter()
+        .map(|p| p.proof.recover_degree_bits(config))
+        .collect::<Vec<_>>();
+    let mut ctl_zs_openings = proofs
+        .iter()
+        .map(|p| p.proof.openings.ctl_zs_last.iter())
+        .collect::<Vec<_>>();
+    for (
+        i,
+        CrossTableLookup {
+            looking_table,
+            looked_table,
+            ..
+        },
+    ) in cross_table_lookups.into_iter().enumerate()
+    {
+        let looking_degree = 1 << degrees_bits[looking_table as usize];
+        let looked_degree = 1 << degrees_bits[looked_table as usize];
+        let looking_z = *ctl_zs_openings[looking_table as usize].next().unwrap();
+        let looked_z = *ctl_zs_openings[looked_table as usize].next().unwrap();
+        ensure!(
+            looking_z
+                == looked_z
+                    * challenges.challenges[i % config.num_challenges]
+                        .gamma
```

Comment:
It looks like the default tuple is the all-zero tuple here? I think this can work, but will lead to a couple complications for tables like Keccak, where `0 -> 0` isn't a valid input-output pair.

- We won't be able to pad the Keccak table to a power of two with `default`, since it won't satisfy constraints (unless we suppress them)
- On CPU rows where Keccak is not performed, we would need constraints to enforce that `default` data is present, to ensure that rows where Keccak is performed can only contain valid (non-`default`) data

So I think it's probably easiest to define a default tuple for each CTL.

Reply:
Oh yeah I initially put zeros with the plan to change it afterwards, but forgot to make the change. I agree it's definitely better to have a default value 👍